llama_stack/providers/remote/inference/groq/groq_utils.py

```diff
@@ -0,0 +1,247 @@
+    ChatCompletionRequest,
+    ChatCompletionResponse,
+    ChatCompletionResponseEvent,
+    ChatCompletionResponseEventType,
+    ChatCompletionResponseStreamChunk,
+    CompletionMessage,
+    Message,
+    Role,
+    StopReason,
+    ToolCall,
+    ToolCallDelta,
+    ToolCallParseStatus,
+    ToolDefinition,
+    ToolParamDefinition,
+    ToolPromptFormat,
+)
+
+
+def convert_chat_completion_request(
+    request: ChatCompletionRequest,
+) -> CompletionCreateParams:
+    """
+    Convert a ChatCompletionRequest to a Groq API-compatible dictionary.
+    Warns client if request contains unsupported features.
+    """
+
+    if request.logprobs:
+        # Groq doesn't support logprobs at the time of writing
+        warnings.warn("logprobs are not supported yet")
+
+    if request.response_format:
+        # Groq's JSON mode is beta at the time of writing
+        warnings.warn("response_format is not supported yet")
+
+    if request.sampling_params.repetition_penalty != 1.0:
+        # groq supports frequency_penalty, but frequency_penalty and sampling_params.repetition_penalty
+        # seem to have different semantics
+        # frequency_penalty defaults to 0 is a float between -2.0 and 2.0
+        # repetition_penalty defaults to 1 and is often set somewhere between 1.0 and 2.0
+        # so we exclude it for now
+        warnings.warn("repetition_penalty is not supported")
+
+    if request.tool_prompt_format != ToolPromptFormat.json:
+        warnings.warn("tool_prompt_format is not used by Groq. Ignoring.")
+
+    return CompletionCreateParams(
+        model=request.model,
+        messages=[_convert_message(message) for message in request.messages],
+        logprobs=None,
+        frequency_penalty=None,
+        stream=request.stream,
+        max_tokens=request.sampling_params.max_tokens or None,
+        temperature=request.sampling_params.temperature,
+        top_p=request.sampling_params.top_p,
+        tools=[_convert_groq_tool_definition(tool) for tool in request.tools or []],
+        tool_choice=request.tool_choice.value if request.tool_choice else None,
+    )
+
+
+def _convert_message(message: Message) -> ChatCompletionMessageParam:
+    if message.role == Role.system.value:
+        return ChatCompletionSystemMessageParam(role="system", content=message.content)
+    elif message.role == Role.user.value:
+        return ChatCompletionUserMessageParam(role="user", content=message.content)
+    elif message.role == Role.assistant.value:
+        return ChatCompletionAssistantMessageParam(
+            role="assistant", content=message.content
+        )
+    else:
+        raise ValueError(f"Invalid message role: {message.role}")
+
+
+def _convert_groq_tool_definition(tool_definition: ToolDefinition) -> dict:
+    # Groq requires a description for function tools
+    if tool_definition.description is None:
+        raise AssertionError("tool_definition.description is required")
```

Comment:
TODO: running into this error on llama-stack-app examples:

<details>
<summary>Trace</summary>
```python
Traceback (most recent call last):
  File "/Users/aidand/dev/llama-stack/llama_stack/distribution/server/server.py", line 158, in sse_generator
    async for item in event_gen:
  File "/Users/aidand/dev/llama-stack/llama_stack/providers/inline/agents/meta_reference/agents.py", line 153, in _create_agent_turn_streaming
    async for event in agent.create_and_execute_turn(request):
  File "/Users/aidand/dev/llama-stack/llama_stack/providers/inline/agents/meta_reference/agent_instance.py", line 179, in create_and_execute_turn
    async for chunk in self.run(
  File "/Users/aidand/dev/llama-stack/llama_stack/providers/inline/agents/meta_reference/agent_instance.py", line 250, in run
    async for res in self._run(
  File "/Users/aidand/dev/llama-stack/llama_stack/providers/inline/agents/meta_reference/agent_instance.py", line 423, in _run
    async for chunk in await self.inference_api.chat_completion(
  File "/Users/aidand/dev/llama-stack/llama_stack/providers/utils/telemetry/trace_protocol.py", line 101, in async_wrapper
    result = await method(self, *args, **kwargs)
  File "/Users/aidand/dev/llama-stack/llama_stack/distribution/routers/routers.py", line 129, in chat_completion
    return (chunk async for chunk in await provider.chat_completion(**params))
  File "/Users/aidand/dev/llama-stack/llama_stack/providers/utils/telemetry/trace_protocol.py", line 101, in async_wrapper
    result = await method(self, *args, **kwargs)
  File "/Users/aidand/dev/llama-stack/llama_stack/providers/remote/inference/groq/groq.py", line 116, in chat_completion
    request = convert_chat_completion_request(
  File "/Users/aidand/dev/llama-stack/llama_stack/providers/remote/inference/groq/groq_utils.py", line 86, in convert_chat_completion_request
    tools=[_convert_groq_tool_definition(tool) for tool in request.tools or []],
  File "/Users/aidand/dev/llama-stack/llama_stack/providers/remote/inference/groq/groq_utils.py", line 86, in <listcomp>
    tools=[_convert_groq_tool_definition(tool) for tool in request.tools or []],
  File "/Users/aidand/dev/llama-stack/llama_stack/providers/remote/inference/groq/groq_utils.py", line 107, in _convert_groq_tool_definition
    raise AssertionError("tool_definition.description is required")
```
</details>